subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -964,7 +965,9 @@ public Configuration copyRecursive(Spec<? super Dependency> dependencySpec) {
     private DefaultConfiguration createCopy(Set<Dependency> dependencies, Set<DependencyConstraint> dependencyConstraints, boolean recursive) {
         DetachedConfigurationsProvider configurationsProvider = new DetachedConfigurationsProvider();
         RootComponentMetadataBuilder rootComponentMetadataBuilder = this.rootComponentMetadataBuilder.withConfigurationsProvider(configurationsProvider);
-        String newName = name + "Copy";
+        copyCount++;
+        String newName = name + "Copy" + (copyCount > 1 ? copyCount : "");
```

Comment:
Could we have problems because of concurrency?
If we do, you can consider something like:
```java
private AtomicInteger copyCount = new AtomicInteger(-1);
private String getNameWithCopySuffix() {
    int count = copyCount.incrementAndGet();
    return count == 0
        ? name
        : name + "Copy" + count;
}
```
and
```
def "copied configuration have unique names"() {
    buildScript """
        configurations {
          conf
        }

        task check {
            doLast {
                assert configurations.conf.copyRecursive().name == 'conf'
                assert configurations.conf.copyRecursive().name == 'confCopy1'
                assert configurations.conf.copyRecursive().name == 'confCopy2'
                assert configurations.conf.copy().name == 'confCopy3'
                assert configurations.conf.copy().name == 'confCopy4'
            }
        }
    """
    expect:
    succeeds ":check"
}
```

Reply:
There are _edge_ cases I think (that I think we're going to deprecate at one point) where this could theoretically happen. 👍  applied your suggestion.